File: android/app/src/main/java/info/bagen/rust/plaoc/microService/core/NativeConnect.kt
package info.bagen.rust.plaoc.microService.core

import info.bagen.rust.plaoc.microService.helper.AdapterManager
import info.bagen.rust.plaoc.microService.ipc.Ipc
import org.http4k.core.Request


typealias ConnectAdapter = suspend (fromMM: MicroModule, toMM: MicroModule, reason: Request) -> Ipc?

val connectAdapterManager = AdapterManager<ConnectAdapter>()


/** 外部程序与内部程序建立链接的方法 */
suspend fun connectMicroModules(fromMM: MicroModule, toMM: MicroModule, reason: Request): Ipc {
    for (connectAdapter in connectAdapterManager.adapters) {
        val ipc = connectAdapter(fromMM, toMM, reason)
        if (ipc != null) {
            return ipc
        }
    }
    throw Exception("no support connect MicroModules, from:${fromMM.mmid} to:${toMM.mmid}")
}

File: toolkit/dweb-core/src/nativeConnect.ts
import type { $PromiseMaybe } from "@dweb-browser/helper/$PromiseMaybe.ts";
import { AdaptersManager } from "@dweb-browser/helper/fun/AdaptersManager.ts";
import type { MicroModule } from "./MicroModule.ts";
import type { Ipc } from "./ipc/ipc.ts";

export type $ConnectAdapter = (fromMM: MicroModule, toMM: MicroModule, reason: Request) => $PromiseMaybe<Ipc | void>;

export const connectAdapterManager = new AdaptersManager<$ConnectAdapter>();

// /** 外部程序与内部程序建立链接的方法 */
// export const connectMicroModules = async (fromMM: MicroModule, toMM: MicroModule, reason: Request) => {
//   for (const connectAdapter of connectAdapterManager.adapters) {
//     const ipc = await connectAdapter(fromMM, toMM, reason);
//     if (ipc != null) {
//       return ipc;
//     }
//   }
//   throw new Error(`no support connect MicroModules, from:${fromMM.mmid} to:${toMM.mmid}`);
// };
